src/Domain/DoctrineDonationRepository.php

```diff
@@ -90,10 +87,22 @@ public function allocateMatchFunds(Donation $donation): string
             $amountNewlyMatched = bcadd($amountNewlyMatched, $newWithdrawalAmount, 2);
         }
 
-        $this->getEntityManager()->flush(); // Flush `$newWithdrawals` if any.
+        try {
+            // Flush `$newWithdrawals` if any and updates to DB copies of CampaignFundings.
+            $this->getEntityManager()->flush();
+        } catch (\Throwable $exception) {
+            $this->matchingAdapter->releaseNewlyAllocatedFunds();
+
+            // Ensure nothing later tries to persist the pending Donation or CampaignFunding changes.
+            $this->getEntityManager()->close();
+            throw new \RuntimeException(
```

Comment:
I think use of `RuntimeException` implies something not worth catching except at the very top level. But might be worth throwing something more specific (and not derived from this class) here so we can catch it in e.g. the action and tell the donor that matching failed and they can try again.